Mathlib/FieldTheory/IsSepClosed.lean

```diff
@@ -0,0 +1,174 @@
+
+theorem of_exists_root (H : ∀ p : k[X], p.Monic → Irreducible p → Separable p → ∃ x, p.eval x = 0) :
+    IsSepClosed k := by
+  refine ⟨fun p ↦ (fun hsep ↦ Or.inr ?_)⟩
+  intro q hq hdvd
+  simp only [map_id] at hdvd 
+  have hlc : IsUnit (leadingCoeff q)⁻¹ := IsUnit.inv <| Ne.isUnit <|
+    leadingCoeff_ne_zero.2 <| Irreducible.ne_zero hq
+  have hsep' : Separable (q * C (leadingCoeff q)⁻¹) :=
+    Separable.mul (Separable.of_dvd hsep hdvd) ((separable_C _).2 hlc)
+    (by simpa only [← isCoprime_mul_unit_right_right (isUnit_C.2 hlc) q 1, one_mul]
+      using isCoprime_one_right (x := q))
+  have hirr' := hq
+  rw [← irreducible_mul_isUnit (isUnit_C.2 hlc)] at hirr'
+  obtain ⟨x, hx⟩ := H (q * C (leadingCoeff q)⁻¹) (monic_mul_leadingCoeff_inv hq.ne_zero) hirr' hsep'
+  exact degree_mul_leadingCoeff_inv q hq.ne_zero ▸ degree_eq_one_of_irreducible_of_root hirr' hx
+
+theorem degree_eq_one_of_irreducible [IsSepClosed k] {p : k[X]}
+    (hp : Irreducible p) (hsep : p.Separable) : p.degree = 1 :=
+  degree_eq_one_of_irreducible_of_splits hp (IsSepClosed.splits_codomain p hsep)
+
+theorem algebraMap_surjective_of_isIntegral {k K : Type _} [Field k] [Ring K] [IsDomain K]
+    [hk : IsSepClosed k] [Algebra k K] [IsSeparable k K] (hf : Algebra.IsIntegral k K) :
+    Function.Surjective (algebraMap k K) := by
+  refine' fun x => ⟨-(minpoly k x).coeff 0, _⟩
+  have hq : (minpoly k x).leadingCoeff = 1 := minpoly.monic (hf x)
+  have hsep : (minpoly k x).Separable := IsSeparable.separable k x
+  have h : (minpoly k x).degree = 1 :=
+    degree_eq_one_of_irreducible k (minpoly.irreducible (hf x)) hsep
+  have : aeval x (minpoly k x) = 0 := minpoly.aeval k x
+  rw [eq_X_add_C_of_degree_eq_one h, hq, C_1, one_mul, aeval_add, aeval_X, aeval_C,
+    add_eq_zero_iff_eq_neg] at this
+  exact (RingHom.map_neg (algebraMap k K) ((minpoly k x).coeff 0)).symm ▸ this.symm
+
+theorem algebraMap_surjective_of_isAlgebraic {k K : Type _} [Field k] [Ring K] [IsDomain K]
+    [IsSepClosed k] [Algebra k K] [IsSeparable k K] (hf : Algebra.IsAlgebraic k K) :
+    Function.Surjective (algebraMap k K) :=
+  algebraMap_surjective_of_isIntegral (Algebra.isAlgebraic_iff_isIntegral.mp hf)
+
+end IsSepClosed
+
+/-- Typeclass for an extension being a separable closure. -/
+class IsSepClosure (R : Type u) (K : Type v) [CommRing R] [Field K] [Algebra R K]
+    [NoZeroSMulDivisors R K] : Prop where
+  sep_closed : IsSepClosed K
+  separable : IsSeparable R K
+
+theorem isSepClosure_iff (K : Type v) [Field K] [Algebra k K] :
+    IsSepClosure k K ↔ IsSepClosed K ∧ IsSeparable k K :=
+  ⟨fun h => ⟨h.1, h.2⟩, fun h => ⟨h.1, h.2⟩⟩
+
+instance (priority := 100) IsSepClosure.normal (R K : Type _) [Field R] [Field K] [Algebra R K]
+    [IsSepClosure R K] : Normal R K :=
+  ⟨fun x => by apply IsIntegral.isAlgebraic; exact IsSepClosure.separable.isIntegral' x,
+    fun x => @IsSepClosed.splits_codomain _ _ _ (IsSepClosure.sep_closed R) _ _ _ (by
```

Comment:
```suggestion
    fun x => @IsSepClosed.splits_codomain _ _ _ _ (IsSepClosure.sep_closed R) _ _ (by
```
If you make my suggestion earlier then this implicit argument needs to be moved to a slightly different place.